src/gathernodeinputhandler.cpp

```diff
@@ -0,0 +1,74 @@
+//*****************************************************************************
+// Copyright 2021 Intel Corporation
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//*****************************************************************************
+#include "gathernodeinputhandler.hpp"
+
+#include "logging.hpp"
+#include "ov_utils.hpp"
+
+namespace ovms {
+
+GatherNodeInputHandler::GatherNodeInputHandler(uint32_t inputsMissingCount, session_id_t shardsCount) :
+    NodeInputHandler(inputsMissingCount) {
+    expectedDependencies *= shardsCount;
+}
+
+void GatherNodeInputHandler::setInput(const std::string& inputName, InferenceEngine::Blob::Ptr& ptr, session_id_t shardId) {
+    auto it = shardsStorage.find(inputName);
+    if (it == shardsStorage.end()) {
+        shard_map_t shardMap{{shardId, ptr}};
+        auto itDidInsertPair = shardsStorage.emplace(inputName, std::move(shardMap));
+        it = itDidInsertPair.first;
+        if (!itDidInsertPair.second) {
+            throw std::runtime_error("Tried to insert the same input twice with the same shard id");  // TODO recoverable error
+        }
+    } else {
+        it->second.emplace(shardId, ptr);
+    }
+}
+
+Status GatherNodeInputHandler::notifyFinishedDependency() {
+    NodeInputHandler::notifyFinishedDependency();
+    if (expectedDependencies > 0) {
+        return StatusCode::OK;
+    }
+    for (auto& [inputName, shardMap] : shardsStorage) {
+        const auto shardsCount = shardMap.size();
+        SPDLOG_LOGGER_DEBUG(dag_executor_logger, "Consolidating: {} shards for input: {}", shardsCount, inputName);
+        auto tensorDesc = shardMap[0]->getTensorDesc();
+        InferenceEngine::SizeVector newDims{1};
```

Comment:
For gathering D1/G1 (D1 in response G1 in request) we do not add new dimension. is it handled here?

Reply:
Since for now there will be no demiltiplexing at the request stage this check is not needed since gathering will always be from demultiplexer from the "inside" of pipeline.